### PR TITLE
create react-play PWA

### LINKED ISSUE #333: ✨ [Feature request]: PWA integration of our ReactPlay app
hi if anyone is willing to build pwa of our reactplay app is strongly welcome. i could also able to help working with workbox and building serviceworker. thanks.

### PR DESCRIPTION
**Pull request by frankiefab100**
# Description

Creates `React-Play` Progressive web app

Fixes #333

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?
 
Passed the Google Lighthouse PWA report

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: public/manifest.json [hunk 1]
```diff
@@ -1,21 +1,11 @@
 {
-  "short_name": "React App",
-  "name": "Create React App Sample",
+  "short_name": "React Play",
+  "name": "React Play PWA",
```
Comment (atapas) [inline public/manifest.json:3]: @frankiefab100 Let's put the name as `ReactPlay`. Short name also the same.

File: public/worker.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
+const CACHE_NAME = "react play";
```
Comment (atapas) [inline public/worker.js:1]: I would put the cache name without a space.

**Review by atapas — CHANGES_REQUESTED** [review]
@frankiefab100 Thanks a lot for making it. I have tested it with preview URL and works well. Given some initial reviews

File: public/worker.js (hunk 2 above)
Comment (frankiefab100) [reply public/worker.js:1]: Okay, I will fix that

**Comment by frankiefab100** [comment]
How does it look now?

**Comment by atapas** [comment]
> How does it look now?

@frankiefab100 On mobile, if you access the preview build, do you see the "Add to Home Screen" Option and does it work?

Here is the preview build with your changes so far. Can you please test and post some screenshots?

https://react-play-93opqordi-atapas.vercel.app/

**Comment by frankiefab100** [comment]
> > How does it look now?
> 
> @frankiefab100 On mobile, if you access the preview build, do you see the "Add to Home Screen" Option and does it work?
> 
> Here is the preview build with your changes so far. Can you please test and post some screenshots?
> 
> https://react-play-93opqordi-atapas.vercel.app/

Yes, it works

Screenshot_20220630-110811
Screenshot_20220630-110806

**Review by atapas — APPROVED** [review]
Looks good to me

**Comment by atapas** [comment]
> > > How does it look now?
> > 
> > 
> > @frankiefab100 On mobile, if you access the preview build, do you see the "Add to Home Screen" Option and does it work?
> > Here is the preview build with your changes so far. Can you please test and post some screenshots?
> > https://react-play-93opqordi-atapas.vercel.app/
> 
> Yes, it works
> 
> Screenshot_20220630-110811 Screenshot_20220630-110806

Looks good to me. Could you please run lighthouse with only PWA option and see all checks are fine?

**Comment by frankiefab100** [comment]
> > > > How does it look now?
> > > 
> > > 
> > > @frankiefab100 On mobile, if you access the preview build, do you see the "Add to Home Screen" Option and does it work?
> > > Here is the preview build with your changes so far. Can you please test and post some screenshots?
> > > https://react-play-93opqordi-atapas.vercel.app/
> > 
> > 
> > Yes, it works
> > Screenshot_20220630-110811 Screenshot_20220630-110806
> 
> Looks good to me. Could you please run lighthouse with only PWA option and see all checks are fine?

Alright

**Review by atapas — CHANGES_REQUESTED** [review]
@frankiefab100 

While you are checking at your end, I see the following issues:

image


And in the console this error

image

**Comment by frankiefab100** [comment]
Here's the Lighthouse PWA report.

The issue is from the logo/icon format

reactplay-lighthouse

**Comment by frankiefab100** [comment]
Here's my console.
I can't tell where that error is from, since it is not from my PR. 

reaactplay console

**Comment by atapas** [comment]
> Here's the Lighthouse PWA report.
> 
> The issue is from the logo/icon format
> 
> reactplay-lighthouse

Thanks for taking care of the console error

> Here's my console. I can't tell where that error is from, since it is not from my PR.
> 
> reaactplay console

Do you mean, you are not seeing the MIME Type error locally?

**Comment by frankiefab100** [comment]
> > Here's the Lighthouse PWA report.
> > The issue is from the logo/icon format
> > reactplay-lighthouse
> 
> Thanks for taking care of the console error
> 
> > Here's my console. I can't tell where that error is from, since it is not from my PR.
> > reaactplay console
> 
> Do you mean, you are not seeing the MIME Type error locally?

Yes, I can't see that error message in the console.

**Comment by atapas** [comment]
> > > Here's the Lighthouse PWA report.
> > > The issue is from the logo/icon format
> > > reactplay-lighthouse
> > 
> > 
> > Thanks for taking care of the console error
> > > Here's my console. I can't tell where that error is from, since it is not from my PR.
> > > reaactplay console
> > 
> > 
> > Do you mean, you are not seeing the MIME Type error locally?
> 
> Yes, I can't see that error message in the console.

Sure thanks for confirming. Can you take care of the two issues we see with lighthouse?

**Comment by atapas** [comment]
> > > > > Here's the Lighthouse PWA report.
> > > > > The issue is from the logo/icon format
> > > > > reactplay-lighthouse
> > > > 
> > > > 
> > > > Thanks for taking care of the console error
> > > > > Here's my console. I can't tell where that error is from, since it is not from my PR.
> > > > > reaactplay console
> > > > 
> > > > 
> > > > Do you mean, you are not seeing the MIME Type error locally?
> > > 
> > > 
> > > Yes, I can't see that error message in the console.
> > 
> > 
> > Sure thanks for confirming. Can you take care of the two issues we see with lighthouse?
> 
> I will be needing the ReactPlay logo of size 512px * 512px (PNG format)

@nirmalkc can you please get one?

**Comment by atapas** [comment]
@frankiefab100 Could you please join our Discord? That way easy to do the file or message exchanges?

https://discord.gg/gAAkgZSJ

**Comment by nirmalkc** [comment]
512x512
Here is the 512 x 512 size image.

**Comment by atapas** [comment]
Thanks @nirmalkc 

@frankiefab100 please take it forward

**Comment by frankiefab100** [comment]
> Thanks @nirmalkc
> 
> @frankiefab100 please take it forward

Alright

**Comment by frankiefab100** [comment]
Sorry, I got sick.

The PWA passed all the light house report now.

File: public/manifest.json [hunk 3]
```diff
@@ -1,19 +1,25 @@
...
       "type": "image/x-icon"
     },
+    {
+      "src": "maskable_icon.png",
+      "sizes": "64x64 32x32 24x24 16x16 512x512 192x192",
+      "type": "image/png",
+      "purpose": "maskable"
+    },
     {
       "src": "logo192.png",
```
Comment (Angryman18) [inline public/manifest.json:17]: make sure this is also our official reactplay icon.

File: public/worker.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
+const CACHE_NAME = "reactplay";
+
+const urlsToCache = ["/", "/completed"];
```
Comment (Angryman18) [inline public/worker.js:3]: what this completed route does?

File: src/ServiceWorkerRegistration.js [hunk 5]
```diff
@@ -0,0 +1,141 @@
+// This optional code is used to register a service worker.
```
Comment (Angryman18) [inline src/ServiceWorkerRegistration.js:1]: what this file is doing? haven't we already registered our service worker?

File: public/worker.js [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+const self = this;
```
Comment (Angryman18) [inline public/worker.js:5]: we can also use window object here.

File: src/index.js [hunk 7]
```diff
@@ -34,6 +35,8 @@ const Index = () => {
 const container = document.getElementById("root");
 createRoot(container).render(<Index />);
 
+ServiceWorkerRegistration.register();
```
Comment (Angryman18) [inline src/index.js:38]: what is the purpose of doing registration again?

File: src/serviceWorker.js [hunk 8]
```diff
@@ -0,0 +1,71 @@
+/* eslint-disable no-restricted-globals */
+
+// This service worker can be customized!
+// See https://developers.google.com/web/tools/workbox/modules
+// for the list of available Workbox modules, or add any other
+// code you'd like.
+// You can also remove this file if you'd prefer not to use a
+// service worker, and the Workbox build step will be skipped.
+
+import { clientsClaim } from "workbox-core";
```
Comment (Angryman18) [inline src/serviceWorker.js:10]: how are you importing from "workbox-core" and other workbox packages. do we have these package installed?

File: src/serviceWorker.js [hunk 9]
```diff
@@ -0,0 +1,71 @@
+/* eslint-disable no-restricted-globals */
```
Comment (Angryman18) [inline src/serviceWorker.js:1]: why are we caching again in this file since you already cached in worker.js?

**Review by Angryman18 — CHANGES_REQUESTED** [review]

File: public/index.html [hunk 10]
```diff
@@ -48,5 +48,30 @@
       To begin the development, run `npm start` or `yarn start`.
       To create a production bundle, use `npm run build` or `yarn build`.
     -->
+
+    <script>
+      if ("serviceWorker" in navigator) {
```
[the remark this answers is not in the document]
Comment (frankiefab100) [reply public/index.html:53]: Sure

File: public/manifest.json (hunk 3 above)
Comment (frankiefab100) [reply public/manifest.json:17]: Alright, any link to the brand assets for the different logo sizes?

**Comment by atapas** [comment]
Hi @frankiefab10, @looking forward to getting this in. When can we start the final review, let us know

**Comment by Angryman18** [comment]
logo is showing default react one and there is console error i got from our latest development.

Screenshot from 2022-07-09 19-49-29

**Comment by frankiefab100** [comment]
> logo is showing default react one and there is console error i got from our latest development.
> 
> Screenshot from 2022-07-09 19-49-29

I'm waiting for the complete logo sets. I asked in discord. 

Atapas, talked about the error earlier but it's not showing in my console. I will have to check again.

**Comment by Angryman18** [comment]
> > logo is showing default react one and there is console error i got from our latest development.
> > Screenshot from 2022-07-09 19-49-29
> 
> I'm waiting for the complete logo sets. I asked in discord.
> 
> Atapas, talked about the error earlier but it's not showing in my console. I will have to check again.

ok. i have added some of my thoughts you can put your view on that.

**Review by frankiefab100 — COMMENTED** [review]
I have made the necessary changes. Please review and let me know.

### The Lighthouse PWA score
react pwa score

**Comment by atapas** [comment]
> I have made the necessary changes. Please review and let me know.
> 
> ### The Lighthouse PWA score
> react pwa score

Super!!!

**Comment by atapas** [comment]
@Angryman18 let's finish the review.

**Comment by Angryman18** [comment]
serviceWorker is not registered.

**Comment by frankiefab100** [comment]
> serviceWorker is not registered.

I think I commented that since it was caching. Let me add it back.

**Comment by Angryman18** [comment]
is that working now or more commits are coming

**Comment by frankiefab100** [comment]
> is that working now or more commits are coming

Yes, please confirm from your end. Any input from you will be welcomed.

**Comment by Angryman18** [comment]
serviceWorker isn't regisered. you can see the deployed one.

**Comment by atapas** [comment]
@Angryman18 I think you tested it an old preview build. I have generated one now. Can you please retest?

**Comment by Angryman18** [comment]
yes i have seen it in the last commit. it is not registered yet also in the deployed one.

**Comment by atapas** [comment]
@frankiefab100  and @Angryman18, May I request you both to get in a call(Our Discord War Room?) to resolve the pending ones. I see it is almost done. If you both can settle on the pending items, great.

**Comment by frankiefab100** [comment]
> @frankiefab100 and @Angryman18, May I request you both to get in a call(Our Discord War Room?) to resolve the pending ones. I see it is almost done. If you both can settle on the pending items, great.

Yeah, I agree with you.

**Comment by atapas** [comment]
@Angryman18 @frankiefab100 

The build fails due to this error:

Attempted import error: 'register' is not exported from './registerServiceWorker' (imported as 'registerServiceWorker').

**Comment by atapas** [comment]
> @Angryman18 @frankiefab100
> 
> The build fails due to this error:
> 
> Attempted import error: 'register' is not exported from './registerServiceWorker' (imported as 'registerServiceWorker').

Please do a yarn build locally to check if the build passes, then push.

**Comment by Angryman18** [comment]
yes. ok i will make necessary changes and push it.

**Comment by Angryman18** [comment]
looks all good to me.

**Comment by frankiefab100** [comment]
> looks all good to me.

Looks good to me too.

**Review by Angryman18 — APPROVED** [review]

**Review by atapas — APPROVED** [review]

**Comment by atapas** [comment]
@all-contributors please add @frankiefab100  for Code

**Pull request merged**